### ISSUE TITLE
HTTPS issues on your website

### ISSUE THREAD

**Issue by rugk** [mailvelope/mailvelope#401]
Your cipher suites on [mailvelope.com](https://www.ssllabs.com/ssltest/analyze.html?d=mailvelope.com&s=95.143.172.170) and [download.mailvelope.com](https://www.ssllabs.com/ssltest/analyze.html?d=download.mailvelope.com) are suboptimal.

In general you should prioritise ECDHE cipher suites over DHE ones. They are faster and more secure. Also Chrome is dropping DHE support.

However there is a more serious issue involved here: If you look at the client tests on SSLLabs (or just look into your local installation when connecting to Mailvelope) you'll see that Firefox will use `TLS_DHE_RSA_WITH_AES_256_CBC_SHA`! That's SHA1 and AES in CBC mode (which is not as good as GCM  mode, because the former is not authenticated encryption). So especially the usage of SHA-1 is ugly as it is going to be [insecure and therefore depreciated](https://konklone.com/post/why-google-is-hurrying-the-web-to-kill-sha-1)*.

**So why does not Firefox use a better cipher?**
If you look at the cipher order, there are two ciphers, which are ordered above the SHA1 cipher. `TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384` is however a cipher with a SHA384 HMAC, which is not supported by Firefox.
And `TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256` is also *not* supported, because Firefox does not support SHA-256 for AES in CBC mode (only GCM mode and the new `CHACHA20_POLY1305` are supported with SHA-256) . An old [Bugzilla ticket for adding these](https://bugzilla.mozilla.org/show_bug.cgi?id=1029179) is currently marked as "wontfix".

So please improve your cipher config. As you can see although SSLLabs shows `A+` they're not optimal.
BTW A nice SSL/TLS testing tool for Firefox is also [this addon](https://github.com/sibiantony/ssleuth).

* The post talks about SHA-1 in SSL/TLS certificates, which is not the case for your server, but the arguments against SHA-1 still apply.

**Comment by toberndo** [comment]
> Your cipher suites on mailvelope.com and download.mailvelope.com are suboptimal.

The cipher suites you are criticizing are only used on [mailvelope.com](https://www.mailvelope.com/). The server at [download.mailvelope.com](https://download.mailvelope.com/) has a different setup.

> However there is a more serious issue involved here: If you look at the client tests on SSLLabs (or
> just look into your local installation when connecting to Mailvelope) you'll see that Firefox will use 
> TLS_DHE_RSA_WITH_AES_256_CBC_SHA! That's SHA1 and AES in CBC mode (which is not as 
> good as GCM mode, because the former is not authenticated encryption). So especially the usage of 
> SHA-1 is ugly as it is going to be insecure and therefore depreciated*.

You are confusing the usage of SHA1 as a signature algorithm for certificates with the usage of SHA1 as an HMAC. `TLS_DHE_RSA_WITH_AES_256_CBC_SHA` uses a SHA1 for HMAC which is [considered secure](http://crypto.stackexchange.com/questions/15382/hmac-sha1-vs-hmac-sha256).

So yes, GCM would be slightly better for Firefox, but I wouldn't call this a "serious issue".

**Comment by rugk** [comment]
> The cipher suites you are criticizing are only used on mailvelope.com. The server at download.mailvelope.com has a different setup.

Right, download.mailvelope.com is okay. So just fix the main site.

> You are confusing the usage of SHA1 as a signature algorithm for certificates with the usage of SHA1 as an HMAC.

That's why I placed this note at the bottom:
> * The post talks about SHA-1 in SSL/TLS certificates, which is not the case for your server, but the arguments against SHA-1 still apply.

> TLS_DHE_RSA_WITH_AES_256_CBC_SHA uses a SHA1 for HMAC which is considered secure.

Yeah, but it is still not nice.

> but I wouldn't call this a "serious issue".

I have not done so. As I said the configs are suboptimal, but as they can be improved quite easily, I highly recommend you to do so.
